### ISSUE TITLE
OpenShot 2.4: ModuleNotFoundError: No module named 'classes'

### ISSUE THREAD

**Issue by ThothK** [OpenShot/openshot-qt#3442]
Hello,

On a FreeBSD 12 system some update broke OpenShot. I believe by changing one of the dependencies.
FreeBSD 12.0
OpenShot 2.4.4.443
Python 3.7.4
Qt 5.13.2_3

When I start OpenShot, the following is displayed:

> PredatorII: ~: openshot-qt    
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/openshot_qt/launch.py", line 47, in <module>
    from classes import info
ModuleNotFoundError: No module named 'classes'

I have been checking for solutions for the past several months. 

The rest of the error message is:

> During handling of the above exception, another exception occurred:

> Traceback (most recent call last):
  File "/usr/local/bin/openshot-qt", line 11, in <module>
    load_entry_point('openshot-qt==2.4.4.dev2', 'gui_scripts', 'openshot-qt')()
  File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 489, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2852, in load_entry_point
    return ep.load()
  File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2443, in load
    return self.resolve()
  File "/usr/local/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2449, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/lib/python3.7/site-packages/openshot_qt/launch.py", line 52, in <module>
    from classes import info
  File "/usr/local/lib/python3.7/site-packages/openshot_qt/classes/info.py", line 30, in <module>
    from PyQt5.QtCore import QDir
ImportError: /usr/local/lib/python3.7/site-packages/PyQt5/QtCore.so: Undefined symbol "_Zpl14QDeadlineTimerx@Qt_5"

What is this telling me?
I interpret the above as I now have the wrong version of Python and the wrong version of Qt. 

What do I have to do?
If I need to compile for the target system, I think I can do that. What I might not be able to do is to make changes to OpenShot. I do not know Python and stopped using Qt ~ 4.5. 

Thanks

**Comment by ferdnyc** [comment]
Ir looks to me as though your PyQt5 installation isn't compatible with the current Qt — it's likely that Qt was updated, which generally requires a rebuild of PyQt5 on top of the new version. Without rebuilding PyQt5 for the current Qt, it'll be looking for symbols from the old version that may not be present in the current Qt, and that appears to be the exact situation here.

I don't think it has anything to do with the Python version, it's that 'Undefined symbol "_Zpl14QDeadlineTimerx@Qt_5"' error coming from PyQt5/QtCore.so.

If you installed PyQt5 using `pip` (which likely triggered an auto-build against the Qt libs installed at the time), the easiest thing to do would probably to forcibly remove it, then reinstall it so that it's built with symbols matching the current Qt.

If PyQt5 was installed by other means (package collection, self-built, etc.), then the advice is pretty much the same: rebuild it for your current Qt 5.13.2 and reinstall.

**Comment by ferdnyc** [comment]
As far as the 'no module named "classes"' error, that's the OpenShot's launcher failing to load modules from its install directory. I suspect that wouldn't have failed if `classes.info` had imported properly. We can see from the second traceback that it's loaded, because it even attempts to import `PyQt5.QtCore.QDir`.

At which point `PyQt5` breaks, so the `from classes import info` fails and causes the `ImportError`. But if PyQt5 is fixed it should solve that problem as well.

**Comment by ThothK** [comment]
Hi,

After updating Qt, the current error message is:
`
PredatorII: ~: openshot-qt
Loaded modules from installed directory: /usr/local/lib/python3.7/site-packages/openshot_qt
Cannot mix incompatible Qt library (version 0x50d00) with this library (version 0x50d02)
Abort trap (core dumped)
`

A roundabout way of checking Qt version:
`
PredatorII: ~: qmake --version
QMake version 3.1
Using Qt version 5.13.2 in /usr/local/lib/qt5
`

After (re)installing what I believe to be the PyQt5 package I receive the same message
`
PredatorII: ~: openshot-qt
Loaded modules from installed directory: /usr/local/lib/python3.7/site-packages/openshot_qt
Cannot mix incompatible Qt library (version 0x50d00) with this library (version 0x50d02)
Abort trap (core dumped)
`

Versions:
PredatorII: ~: python --version
Python 3.7.4
PredatorII: ~: qmake --version
QMake version 3.1
Using Qt version 5.13.2 in /usr/local/lib/qt5
PredatorII: ~: pkg info py37-qt5
py37-qt5-5.13.1_1

This `py37-qt5-5.13.1_1`  supports the concept of PyQt being the wrong version. I guess it would need to be 5.13.2x.
Does that sound logical and / or correct?


Thank You

**Comment by ferdnyc** [comment]
Ah! Looks like the issue now is, because _libopenshot_ is also linked with Qt libraries, it too has to be recompiled with the new version. From the error it appears your current libopenshot was built with Qt 5.13.0.

> This `py37-qt5-5.13.1_1` supports the concept of PyQt being the wrong version. I guess it would need to be 5.13.2x.

Believe it or not, it doesn't. PyQt5 has its own versioning, and while they tend to release a new version with each Qt release, the PyQt5 version _doesn't_ HAVE to be the same as the Qt version. So, what's **illogical** (but sadly correct) is that you can have a PyQt5 version 5.13.1 that uses Qt 5.13.2, and there's no actual problem with that.

And it appears that's the case, because the error message you're getting is showing a mismatch between Qt version `5.13.0` (`0x50d00`) and version `5.13.2` (`0x50d02`), neither of which is `5.13.1`.

So, it still appears to me that your PyQt5 is now fine (it's PyQt5 5.13.1 linked with Qt 5.13.2), but it's most likely libopenshot that's picking up Qt `5.13.0`.

Or I suppose it's possible that your PyQt5 5.13.1 is linked with Qt 5.13.2, but if you built it yourself that's unlikely. If you want to check, the best approach is to ask it directly:
```python3
>>> from PyQt5 import Qt
>>> print(Qt.PYQT_VERSION_STR)
5.13.2
>>> print(Qt.PYQT_VERSION_STR)
5.13.2
```
We don't export the version of Qt we build with in libopenshot's Python bindings anywhere (though that's probably not a bad idea), so the same thing won't work there. And in fact, since we don't do that, there doesn't appear to be an easy way to determine it. The symbols in the library only contain identifiers like `Qt_5` and `Qt_5.13`, nothing more specific than that.

But if your PyQt5 shows a `QT_VERSION_STR` of 5.13.2, then you can probably assume that libopenshot is the problem, especially if there's any chance it didn't get rebuilt after updating to 5.13.2.

**Comment by ferdnyc** [comment]
> Or I suppose it's possible that your PyQt5 5.13.1 is linked with Qt 5.13.2, but if you built it yourself that's unlikely.

I meant "is linked with Qt 5.13.0", there.

**Comment by ferdnyc** [comment]
> ```python
> >>> from PyQt5 import Qt
> >>> print(Qt.PYQT_VERSION_STR)
> 5.13.2
> >>> print(Qt.PYQT_VERSION_STR)
> 5.13.2
> ```

Aargh! And I meant `Qt.QT_VERSION_STR`, for the first one of those:

> ```python
> >>> from PyQt5 import Qt
> >>> print(Qt.QT_VERSION_STR)
> 5.13.2
> >>> print(Qt.PYQT_VERSION_STR)
> 5.13.2
> ```

**Comment by ferdnyc** [comment]
I just submitted OpenShot/libopenshot#510 adding `openshot.QT_VERSION_STR` to libopenshot (including the Python bindings). That should make it easier to examine the library config when necessary, in the future.

**Comment by ThothK** [comment]
Hello Frank,

The output of the interactive check you described is:
`
PredatorII: ~: python3
Python 3.7.4 (default, Aug 22 2019, 05:59:46) 
[Clang 6.0.1 (tags/RELEASE_601/final 335540)] on freebsd12
Type "help", "copyright", "credits" or "license" for more information.
>>> from PyQt5 import Qt
>>> print(Qt.QT_VERSION_STR)
5.13.2
>>> print(Qt.PYQT_VERSION_STR)
5.13.1
>>> exit() 
`
I did not build the ports myself. Instead I chose to rely on the package system on FreeBSD. 

>  So, what's illogical (but sadly correct) is that you can have a PyQt5 version 5.13.1 that uses Qt 5.13.2, and there's no actual problem with that.
If I understand correctly, there appears to be a problem in this case. 

Am I correct in thinking that I need to contact the maintainer and ask him / her to rebuild ... or at least relink against 5.13.2?


Thank You

**Comment by ThothK** [comment]
Reopened as I mistakenly clicked the close and comment button.

**Comment by ferdnyc** [comment]
@ThothK 

Maintainer of which? your PyQt/Qt situation looks fine, and exactly what I expected: PyQt 5.13.1 linked with Qt 5.13.2. Confusing and a bit nonsensical, but perfectly valid.

Which leads me to the conclusion that libopenshot is causing the version mismatches, because it was most likely originally built with Qt 5.13.0. So, libopenshot probably needs to be rebuilt, by... whoever built it. If there's a libopenshot maintainer, then it likely falls to them. They may not have realized that the Qt components mean it needs to be rebuilt every time Qt is updated, to avoid conflicts with packages using the newer Qt.

**Comment by ThothK** [comment]
Hello,

After speaking with the FreeBSD maintainer of OpenShot I was able to find what appears to be the source of the problem I was experiencing.

For whatever reason, the tool used to manage pre-compiled programs did not fully update Qt. I had components that were marked as v5.13.0 and others that were marked as 5.13.2.
![image](https://user-images.githubusercontent.com/5565436/82436325-7d612f80-9a4a-11ea-9cc1-9651c302ef8c.png)

In the end, I manually updated the components that were marked as 5.13.0. Afterwards, OpenShot displayed a UI after starting.

Thank you for the effort in helping me understand what was happening.
